### ISSUE TITLE
Missed report of signed overflow (signed-integer-overflow)

### ISSUE THREAD

**Issue by JonPsson** [llvm/llvm-project#55480]
```c
int printf ( const char * format, ... );
int a = -1181091070, b, c;
short e;

int f(int g, int h) {
  return g > 2147483647 - h || h < g < -h ? g : g + h;
}

int main() {
  int i = a;
  b = f(i, &e != 0);
  c = i > 0 && b > 2147483647 - i || b - i ? b : i;
  printf("%d\n", c);
}

```
The above program contains a signed subtraction which will overflow (2147483647 - i, which is SINT32_MAX plus something), but the sanitizer does not report this. I compiled with:

```shell
clang -O3  -march=arch13 -o -  -fsanitize=undefined
```

and see in the output assembly:

```
.LBB1_4:                                # %lor.lhs.false
        srk     %r12, %r0, %r13
        jo      .LBB1_7
...
.LBB1_7:                                # %handler.sub_overflow2
        llgfr   %r3, %r0
        larl    %r2, .L__unnamed_4
        lgr     %r4, %r13
        brasl   %r14, __ubsan_handle_sub_overflow@PLT
```
I suspect this is a "Jump if Ordered" to the block where the sanitizer handles the sub overflow, but since the SRK actually overflows it does not go there. So the branch looks to me like having the wrong condition.

Before ISel I see:

```llvm
land.lhs.true:                                    ; preds = %f.exit.thread, %lor.lhs.false.i
  %cond.i11 = phi i32 [ 2147483647, %f.exit.thread ], [ %4, %lor.lhs.false.i ]
  %5 = trunc i64 %1 to i32
  %6 = sub nuw nsw i32 2147483647, %5
  %cmp1 = icmp ugt i32 %cond.i11, %6
  br i1 %cmp1, label %cond.end, label %lor.lhs.false

lor.lhs.false:                                    ; preds = %land.lhs.true, %lor.lhs.false.i
  %7 = phi i32 [ %cond.i11, %land.lhs.true ], [ %4, %lor.lhs.false.i ]
  %8 = trunc i64 %1 to i32
  %9 = tail call { i32, i1 } @llvm.ssub.with.overflow.i32(i32 %7, i32 %8), !nosanitize !4
  %10 = extractvalue { i32, i1 } %9, 0, !nosanitize !4
  %11 = extractvalue { i32, i1 } %9, 1, !nosanitize !4
  br i1 %11, label %handler.sub_overflow2, label %cont3, !prof !5, !nosanitize !4
```
%5 is the loaded value of a (-1181091070). The icmp is skipping the sanitizer block if SINT32_MAX is greater than the result of the sub, which it is. This looks like the needed check is missed due to a broken condition around it, or?

**Comment by nikic** [comment]
`i > 0` is already false, so `b > 2147483647 - i` never gets executed, unless I'm misreading the code?

**Comment by JonPsson** [comment]
> `i > 0` is already false, so `b > 2147483647 - i` never gets executed, unless I'm misreading the code?

Hmm, yes it looks like that in the C code, but in the optimized code this subtraction overflow was indeed executed *without* the sanitizer enabled. So maybe this is just a case where the sanitizer changes the optimization results and therefore are not effective on the code originally run?
